src/Mod/Sketcher/App/planegcs/Constraints.cpp

```diff
@@ -129,6 +129,171 @@ double ConstraintEqual::grad(double *param)
     return scale * deriv;
 }
 
+// Weighted Linear Combination
+
+const std::vector< std::vector<double> > ConstraintWeightedLinearCombination::defaultfactors =
+{{},
+ {1.0},
+ {0.5, 0.5},
+ {0.16666666666666666, 0.6666666666666667, 0.16666666666666669},
+ {0.041666666666666664, 0.45833333333333337, 0.4583333333333333, 0.04166666666666667},
+ {0.008333333333333333, 0.21666666666666667, 0.55, 0.2166666666666667, 0.008333333333333333},
+ {0.0013888888888888887, 0.07916666666666666, 0.41944444444444445, 0.41944444444444445, 0.07916666666666669, 0.0013888888888888885},
+ {0.0001984126984126984, 0.023809523809523815, 0.2363095238095238, 0.4793650793650793, 0.2363095238095238, 0.02380952380952381, 0.0001984126984126984},
+ {2.48015873015873e-05, 0.006125992063492064, 0.10647321428571428, 0.3873759920634921, 0.38737599206349216, 0.10647321428571431, 0.006125992063492064, 2.48015873015873e-05},
+ {2.7557319223985893e-06, 0.0013833774250440916, 0.04025573192239859, 0.24314925044091712, 0.43041776895943573, 0.24314925044091715, 0.04025573192239859, 0.0013833774250440916, 2.7557319223985905e-06}};
+
+
+ConstraintWeightedLinearCombination::ConstraintWeightedLinearCombination(size_t givennumpoints, const std::vector<double *>& givenpvec)
+    : numpoints(givennumpoints)
+{
+    pvec = givenpvec;
+    setupInputs();
+}
+
+ConstraintWeightedLinearCombination::ConstraintWeightedLinearCombination(size_t givennumpoints, const std::vector<double *>& givenpvec, const std::vector<double>& givenfactors)
+    : factors(givenfactors)
+    , numpoints(givennumpoints)
+{
+    pvec = givenpvec;
+    setupInputs();
+}
+
+void ConstraintWeightedLinearCombination::setupInputs()
+{
+    assert(pvec.size() == 2*numpoints + 1);
+    if (factors.empty())
+        factors = defaultfactors[numpoints];
+    assert(factors.size() == numpoints);
+    origpvec = pvec;
+    rescale();
+}
+
+ConstraintType ConstraintWeightedLinearCombination::getTypeId()
+{
+    return WeightedLinearCombination;
+}
+
+void ConstraintWeightedLinearCombination::rescale(double coef)
+{
+    scale = coef * 1.;
+}
+
+double ConstraintWeightedLinearCombination::error()
+{
+    double offset = 1;
+    double woffset = 1 + numpoints;
+
+    double sum = 0;
+    double wsum = 0;
+
+    for (size_t i = 0; i < numpoints; ++i) {
+        double wcontrib = *pvec[woffset + i] * factors[i];
+        wsum += wcontrib;
+        sum += *pvec[offset + i] * wcontrib;
+    }
+
+    return scale * ((*pvec[0]) * wsum - sum);
+}
+
+double ConstraintWeightedLinearCombination::grad(double *param)
+{
+    double offset = 1;
+    double woffset = 1 + numpoints;
+
+    // TODO: Do we assume pvec are different so param can only be one?
+
+    double deriv=0.;
+    if (param == pvec[0]) {
+        double wsum = 0;
+        for (size_t i = 0; i < numpoints; ++i) {
+            wsum += *pvec[woffset + i] * factors[i];
+        }
+        deriv += wsum;
+    }
+
+    for (size_t i = 0; i < numpoints; ++i) {
```

Comment:
Ok. This loop is intended to find the actual pole (not as a summatory like the one above), see also equations (12) and (13) of your PDF.

Please, add as a comment to each block the equation you are implementing in your pdf for clarity (also to the one of the partial with respect to the point above).

Then there are several issues, some of efficiency and others of intention. The latter are the most important to me.

deriv += is misleading and should be deriv =, as a single element contributes to the derivative (there is no sum of different contributions in the derivative).

A soon as we have found both derivatives, we can return (we do not need to run the full loop). Efficiency aside (which would be negligible given the size and today's computers), the intention is important, because this looks what is not. I would define two booleans found_partial_p, found_partial_w return scale*deriv when both have been found.

Reply:
> Please, add as a comment to each block the equation you are implementing in your pdf for clarity (also to the one of the partial with respect to the point above).

The only problem here is that the document is not exactly an archive. In the future if I wish to add general tangent to B-spline or variable knot parameters the equation numbers may change.
 
> Then there are several issues, some of efficiency and others of intention. The latter are the most important to me.

For the intention issues, could you look at, for example, the tangent-at-knot constraint?
 
> deriv += is misleading and should be deriv =, as a single element contributes to the derivative (there is no sum of different contributions in the derivative).

Again, this is taken from other constraints.

> A soon as we have found both derivatives, we can return (we do not need to run the full loop). Efficiency aside (which would be negligible given the size and today's computers), the intention is important, because this looks what is not. I would define two booleans found_partial_p, found_partial_w return scale*deriv when both have been found.

Note that at any point only found_partial_p or found_partial_w can be true. A parameter can only be a pole value (here the x or y coordinate) or the pole weight.

Reply:
> The only problem here is that the document is not exactly an archive. In the future if I wish to add general tangent to B-spline or variable knot parameters the equation numbers may change.

For now this will have to do. I'll have to be careful of not changing the linked document and somehow versioning future editions.

Reply:
Yes. I did not realise at that moment that can only be one found. The amended you just pushed is ok with me. Thanks.

It is not convenient, but it is not the end of the world if the equation gets renumbered, the developer can really "correct" the equation number and find the right one, once he is indicated that there is a specific document with one equation.